File: 정종길/프로그래머스/양과 늑대.py
from collections import deque


def solution(info, edges):
    global answer, graph
    n = len(info)
    answer = 0
    graph = [[] for _ in range(n)]

    for p, c in edges:
        graph[p].append(c)

    bfs(info)

    return answer


def bfs(info):
    global answer, graph

    q = deque([(0, 1, 0, graph[0])])

    while q:
        now, sheep, wolf, next_ = q.popleft()
        for i in next_:
            if info[i]:
                if wolf + 1 >= sheep:
                    continue
                else:
                    nnext_ = next_[:]
                    nnext_.remove(i)
                    q.append((i, sheep, wolf + 1, nnext_ + graph[i]))
            else:
                nnext_ = next_[:]
                nnext_.remove(i)
                q.append((i, sheep + 1, wolf, nnext_ + graph[i]))

        answer = max(answer, sheep)

File: 정종길/프로그래머스/카카오 1회독/2022 KAKAO BLIND RECRUITMENT/양과 늑대.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int[] info, int[][] edges) {
        int answer = 0;
        List<ArrayList<Integer>> graph = IntStream.range(0,18).mapToObj(i -> new ArrayList<Integer>()).collect(Collectors.toList());

        for(int [] edge : edges){
            graph.get(edge[0]).add(edge[1]);
        }

        Queue<Object[]> q = new LinkedList<>();

        q.add(new Object[]{0, 1, 0, new ArrayList<>(graph.get(0))});

        while(!q.isEmpty()){
            Object[] data = q.poll();
            int now = (int) data[0];
            int sheep = (int) data[1];
            int wolf = (int) data[2];
            ArrayList<Integer> nextList = (ArrayList<Integer>) data[3];

            if(sheep > answer){
                answer = sheep;
            }

            for(int next : nextList){
                if(info[next] == 1){
                    if(sheep - (wolf + 1) > 0){
                        ArrayList<Integer> nnextList = new ArrayList<>(nextList);
                        nnextList.remove(Integer.valueOf(next));
                        nnextList.addAll(new ArrayList(graph.get(next)));
                        q.add(new Object[]{ next, sheep, wolf + 1, nnextList});
                    }

                } else {
                    ArrayList<Integer> nnextList = new ArrayList<>(nextList);
                    nnextList.remove(Integer.valueOf(next));
                    nnextList.addAll(new ArrayList(graph.get(next)));
                    q.add(new Object[]{ next, sheep + 1, wolf, nnextList});
                }
            }
        }
        return answer;
    }
}